pkg/network/port_test.go

```diff
@@ -52,141 +50,136 @@ func TestMain(m *testing.M) {
 	os.Exit(m.Run())
 }
 
-func TestReadInitialTCPState(t *testing.T) {
-	nsName := netlinktestutil.AddNS(t)
-	t.Cleanup(func() {
-		err := exec.Command("testdata/teardown_netns.sh").Run()
-		assert.NoError(t, err, "failed to teardown netns")
-	})
-
-	err := exec.Command("testdata/setup_netns.sh", nsName).Run()
-	require.NoError(t, err, "setup_netns.sh failed")
-
-	l, err := net.Listen("tcp", ":0")
-	require.NoError(t, err)
-	defer func() { _ = l.Close() }()
-
-	l6, err := net.Listen("tcp6", ":0")
-	require.NoError(t, err)
-	defer func() { _ = l.Close() }()
+var (
+	ip4Re = regexp.MustCompile(`.+listening on.+0.0.0.0:([0-9]+)`)
+	ip6Re = regexp.MustCompile(`.+listening on.+\[0000:0000:0000:0000:0000:0000:0000:0000\]:([0-9]+)`)
+)
 
-	ports := []uint16{
-		getPort(t, l),
-		getPort(t, l6),
-		34567,
-		34568,
+// runServerProcess runs a server using `socat` externally
+//
+// `proto` can be "tcp4", "tcp6", "udp4", or "udp6"
+// `port` can be `0` in which case the os assigned port is returned
+func runServerProcess(t *testing.T, proto string, port uint16, ns netns.NsHandle) uint16 {
+	var re *regexp.Regexp
+	address := fmt.Sprintf("%s-listen:%d", proto, port)
+	switch proto {
+	case "tcp4", "udp4":
+		re = ip4Re
+	case "tcp6", "udp6":
+		re = ip6Re
+	default:
+		require.FailNow(t, "unrecognized protocol")
 	}
 
-	ns, err := netns.GetFromName(nsName)
-	require.NoError(t, err)
-	defer ns.Close()
-
-	nsIno, err := kernel.GetInoForNs(ns)
-	require.NoError(t, err)
-
-	require.Eventually(t, func() bool {
-		initialPorts, err := ReadInitialState("/proc", TCP, true)
-		require.NoError(t, err)
-		for _, p := range ports[:2] {
-			if _, ok := initialPorts[PortMapping{testRootNs, p}]; !ok {
-				t.Errorf("PortMapping(testRootNs) returned false for port %d", p)
-				return false
-			}
+	kernel.WithNS(ns, func() error {
+		cmd := exec.Command("socat", "-d", "-d", "STDIO", address)
+		stderr, err := cmd.StderrPipe()
+		require.NoError(t, err, "error getting stderr pipe for command %s", cmd)
+		require.NoError(t, cmd.Start())
+		t.Cleanup(func() { cmd.Process.Kill() })
+		if port != 0 {
+			return nil
 		}
-		for _, p := range ports[2:] {
-			if _, ok := initialPorts[PortMapping{nsIno, p}]; !ok {
-				t.Errorf("PortMapping(test ns) returned false for port %d", p)
-				return false
+
+		scanner := bufio.NewScanner(stderr)
```

Comment:
so if we pass a `0` port to socat it returns the port it assigned to `sterr`?

Reply:
In its logs, yes.

Reply:
that is weird and interesting to me